dhis-2/dhis-services/dhis-service-expression-parser/src/main/java/org/hisp/dhis/expressionparser/ExpressionVisitor.java

```diff
@@ -0,0 +1,596 @@
+
+        return double2( d1, d2, fn);
+    }
+
+    /**
+     * Evaluates a two-argument double function, but returns null if either of
+     * the input values is null.
+     *
+     * @param d1 the first function argument (could be null).
+     * @param d2 the second function argument (could be null).
+     * @param fn the function to evaluate.
+     * @return the function result, or null if either input was null.
+     */
+    protected Double double2( Double d1, Double d2, BiFunction<Double, Double, Double> fn )
+    {
+        return d1 == null || d2 == null
+            ? null
+            : fn.apply( d1, d2 );
+    }
+
+    /**
+     * Evaluates a two-argument boolean function, but returns null if either of
+     * the input values is null.
+     *
+     * @param ctx the parsing context with two expressions.
+     * @param fn the function to evaluate.
+     * @return the function result, or null if either input was null.
+     */
+    protected Boolean boolean1( ExprContext ctx, Function<Boolean, Boolean> fn )
+    {
+        Boolean b1 = castBoolean( visit( ctx.expr( 0 ) ) );
+
+        return b1 == null
+            ? null
+            : fn.apply( b1 );
+    }
+
+    /**
+     * Casts object as Integer, or throws an exception if we can't.
+     *
+     * @param object
+     * @return
+     */
+    protected Integer castInteger( Object object )
+    {
+        Double d = castDouble( object );
+
+        Integer i = null;
+
+        if ( d != null )
+        {
+            i = (int) (double) d;
+
+            if ( (double) d != i )
+            {
+                throw new ExpressionParserExceptionWithoutContext( "Integer expected" );
+            }
+        }
+
+        return i;
+    }
+
+    /**
+     * Casts object as Double, or throws exception.
+     * <p/>
+     * If the object is null, return null.
+     *
+     * @param object the value to cast as a Double.
+     * @return Double value.
+     */
+    protected Double castDouble( Object object )
+    {
+        return (Double) cast( Double.class, object );
+    }
+
+    /**
+     * Casts object as Boolean, or throws exception.
+     *
+     * @param object the value to cast as a Boolean.
+     * @return Boolean value.
+     */
+    protected Boolean castBoolean( Object object )
+    {
+        return (Boolean) cast( Boolean.class, object );
+    }
+
+    /**
+     * Casts object as String, or throws exception.
+     *
+     * @param object the value to cast as a String.
+     * @return String value.
+     */
+    protected String castString( Object object )
+    {
+        return (String) cast( String.class, object );
+    }
+
+    /**
+     * Checks to see whether object can be cast to the class specified,
+     * or throws exception if it can't.
+     *
+     * @param clazz the class: Double, Boolean, or String
+     * @param object the value to cast
+     * @return object (if it can be cast to that class.)
+     */
+    protected Object cast( Class clazz, Object object )
```

Comment:
Class is a generic class so should have at least `Class<?>` generic type

Reply:
Done.